File: internal/gitkit/ts/bin/github-io.ts
// github-io.ts — mechanical GitHub I/O the skills call. Zero npm deps.
// Run: node github-io.ts <subcommand> --key value ...
// Subcommands: reply, resolve-thread, comment, react, review, create-pr,
//              detect-workflows, find-run, watch-run, failed-logs, rerun-failed

import { execFileSync } from "node:child_process";
import { readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";

import { repoRoot } from "./repo-root.ts";

const RESOLVE_THREAD_MUTATION =
  "mutation($threadId:ID!){ resolveReviewThread(input:{threadId:$threadId}){ thread { isResolved } } }";

// Running the script bare is how a caller discovers its interface. Until
// 2026-08-08 that answered `Unknown github-io subcommand: undefined`, which
// teaches nothing — hence this synopsis.
const USAGE = `github-io — mechanical GitHub I/O for the git skills.

Usage: node github-io.ts <subcommand> --key value ...

  reply             --owner O --repo R --pr N --commentId ID --body TEXT
  comment           --owner O --repo R --pr N --body TEXT
  react             --owner O --repo R --commentId ID [--content +1]
  review            --owner O --repo R --pr N --event request-changes|comment --body TEXT
  resolve-thread    --threadId PRRT_...
  create-pr         --head BRANCH --base BRANCH [--title T] [--body B] [--draft] [--label L]
  find-run          --sha SHA
  watch-run         --runId ID
  failed-logs       --runId ID
  rerun-failed      --runId ID
  detect-workflows  (no flags — scans ./.github/workflows)

Flags are camelCase (--commentId, --threadId, --runId); owner/repo are never
inferred from the cwd. \`review\` cannot approve, by design.

The harness shell is zsh, which does NOT word-split unquoted \$VAR — packing
flags into one variable sends them as a SINGLE argument. Use an array:
  FLAGS=(--owner o --repo r --pr 1); node github-io.ts reply "\${FLAGS[@]}" --commentId 9 --body X
`;

function req(o: Record<string, string>, key: string): string {
  const v = o[key];
  if (v == null || v === "") throw new Error(`missing required field: ${key}`);
  return v;
}

export function buildCommand(sub: string, o: Record<string, string>): string[] {
  switch (sub) {
    case "find-run":
      return ["run", "list", "--commit", req(o, "sha"),
        "--json", "databaseId,status,conclusion,workflowName,headSha", "--limit", "20"];
    case "watch-run":
      return ["run", "watch", req(o, "runId"), "--exit-status"];
    case "failed-logs":
      return ["run", "view", req(o, "runId"), "--log-failed"];
    case "rerun-failed":
      return ["run", "rerun", req(o, "runId"), "--failed"];
    case "reply":
      return ["api", "--method", "POST",
        `repos/${req(o, "owner")}/${req(o, "repo")}/pulls/${req(o, "pr")}/comments/${req(o, "commentId")}/replies`,
        "-f", `body=${req(o, "body")}`];
    case "resolve-thread":
      return ["api", "graphql", "-f", `query=${RESOLVE_THREAD_MUTATION}`, "-f", `threadId=${req(o, "threadId")}`];
    // ── Closure for the NON-THREAD surfaces ────────────────────────────────
    // A review's summary body and a PR conversation comment have no review
    // thread, so `resolve-thread` cannot touch them and `reply` (which posts into
    // a review thread) does not apply either. GitHub gives them exactly two
    // affordances, and together they are what "handled" means for those surfaces:
    case "comment":
      // A PR-level comment. Quote what you are answering — this lands at the
      // bottom of the conversation, not under the comment it addresses.
      return ["api", "--method", "POST",
        `repos/${req(o, "owner")}/${req(o, "repo")}/issues/${req(o, "pr")}/comments`,
        "-f", `body=${req(o, "body")}`];
    case "react":
      // An idempotent, non-noisy "seen/actioned" marker on a conversation
      // comment. GitHub exposes reactions for ISSUE comments only — there is no
      // reactions endpoint for a review's summary body, so a review summary is
      // closed by the quoting `comment` alone.
      return ["api", "--method", "POST",
        `repos/${req(o, "owner")}/${req(o, "repo")}/issues/comments/${req(o, "commentId")}/reactions`,
        "-f", `content=${o.content || "+1"}`];
    case "review": {
      // A verdict-carrying PR review — what `comment` structurally cannot be. Needed by
      // the `--auto` regression audit (`auto-audit.md`): a plain conversation
      // comment is invisible to an autonomous PR author, whereas CHANGES_REQUESTED is the
      // one signal eve peacemaker's passive review loop (D6) matches to its claim and
      // answers with a revision. Blocking a machine-authored PR with a comment alone
      // means nothing happens; blocking it with a review makes the author fix it.
      //
      // APPROVE IS DELIBERATELY UNREACHABLE. "Never self-approve" is a hard rule in
      // merge/SKILL.md, and `--auto` removes the human who used to be the one enforcing
      // it. Encoding the ban in the argv builder means no caller can reach it by
      // forgetting — the affordance does not exist, rather than existing and being
      // discouraged.
      const event = req(o, "event");
      if (event !== "request-changes" && event !== "comment") {
        throw new Error(`review event must be request-changes|comment (never approve): got ${event}`);
      }
      return ["pr", "review", req(o, "pr"), `--${event}`, "--body", req(o, "body"),
        "--repo", `${req(o, "owner")}/${req(o, "repo")}`];
    }
    case "create-pr":
      // Ready-for-review by default (no --draft) per the global PR-defaults rule;
      // --draft is opt-in only (callers pass it explicitly to override the default).
      //
      // --fill stays as the FALLBACK (title/body from commits). gh lets --title
      // and --body sit alongside it and gives them precedence, so passing both is
      // correct: an explicit title wins, and anything not supplied is still
      // autofilled. Forwarding these was missing until 2026-07-26 — buildCommand
      // rebuilt the argv from scratch, so a caller's --title/--body never reached
      // gh at all and the PR silently got a commit-derived body.
      return ["pr", "create", "--head", req(o, "head"), "--base", req(o, "base"), "--fill",
        ...(o.title ? ["--title", o.title] : []),
        ...(o.body ? ["--body", o.body] : []),
        ...("draft" in o ? ["--draft"] : []),
        // MERGE_POLICY=self repos: `eve-ignore` keeps eve off a PR nobody will wait on.
        ...(o.label ? ["--label", o.label] : [])];
    default:
      throw new Error(`Unknown github-io subcommand: ${sub}`);
  }
}

// detect-workflows is filesystem-based, not gh-backed.
export function detectWorkflows(cwd: string): { name: string; path: string }[] {
  const dir = join(cwd, ".github", "workflows");
  let files: string[] = [];
  try {
    files = readdirSync(dir).filter((f) => f.endsWith(".yml") || f.endsWith(".yaml"));
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") return []; // no workflows dir → empty, not an error
    throw e;
  }
  return files.map((f) => {
    const text = readFileSync(join(dir, f), "utf8");
    const m = text.match(/^name:\s*(.+)$/m);
    return { name: m ? m[1].trim().replace(/^['"]|['"]$/g, "") : f, path: `.github/workflows/${f}` };
  });
}

export function parseFlags(argv: string[]): Record<string, string> {
  const o: Record<string, string> = {};
  for (let i = 0; i < argv.length; i++) {
    if (!argv[i].startsWith("--")) continue;
    const key = argv[i].slice(2);
    // A key with whitespace means several flags arrived glued into ONE argv
    // token. Overwhelmingly that is zsh (the harness shell) not word-splitting
    // an unquoted $VAR: `O="--owner o --repo r"; node github-io.ts reply $O`
    // passes one argument, every real flag is lost, and the old failure mode was
    // a baffling `missing required field: owner` when owner was right there
    // (2026-08-08, PR #1066 reply round). Name the real cause instead.
    if (/\s/.test(key)) {
      throw new Error(
        `flag arrived as ONE argument with embedded spaces: "--${key}"\n` +
        `  The harness shell is zsh, which does NOT word-split unquoted $VAR.\n` +
        `  Pass the flags literally, or use an array:\n` +
        `    FLAGS=(--owner o --repo r --pr 1); node github-io.ts <sub> "\${FLAGS[@]}"`,
      );
    }
    const next = argv[i + 1];
    // A valueless flag (next token is another flag, or there is none) records "".
    // Without this, `--draft --title X` would swallow "--title" as draft's value —
    // which is why callers used to be told to pass --draft last. Now order is free.
    o[key] = next != null && !next.startsWith("--") ? argv[++i] : "";
  }
  return o;
}

async function main(): Promise<void> {
  const [sub, ...rest] = process.argv.slice(2);
  // Explicit help → stdout, exit 0. A bare invocation still exits 1, so a script
  // whose `$SUB` expanded to nothing fails loudly instead of looking successful.
  if (sub === "help" || sub === "--help" || sub === "-h") {
    process.stdout.write(USAGE);
    return;
  }
  if (sub === undefined) {
    process.stderr.write(USAGE);
    process.exit(1);
  }
  if (sub === "detect-workflows") {
    process.stdout.write(JSON.stringify(detectWorkflows(process.cwd()), null, 2) + "\n");
    return;
  }
  const argv = buildCommand(sub, parseFlags(rest));
  // repoRoot([]) — NEVER let it sniff process.argv: github-io's own `--repo`
  // is a GitHub repo NAME (owner/repo split across flags), not a directory
  // anchor, and the default argv scan misreads it and throws. GIT_SKILL_REPO
  // env anchoring still applies; otherwise the ambient cwd's toplevel.
  const out = execFileSync("gh", argv, { encoding: "utf8", stdio: ["ignore", "pipe", "inherit"], maxBuffer: 64 * 1024 * 1024, timeout: 60_000, cwd: repoRoot([]) });
  process.stdout.write(out);
}

if (import.meta.main) {
  main().catch((e) => {
    process.stderr.write(`error: ${(e as Error).message}\n`);
    process.exit(1);
  });
}

File: internal/gitkit/githubio.go
package gitkit

import (
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"syscall"
	"time"
)

// github-io — mechanical GitHub I/O the skills call:
// `github-io <subcommand> --key value ...`.

const resolveThreadMutation = "mutation($threadId:ID!){ resolveReviewThread(input:{threadId:$threadId}){ thread { isResolved } } }"

// githubIOUsage answers a bare or help invocation: running the verb bare is
// how a caller discovers its interface.
const githubIOUsage = `github-io — mechanical GitHub I/O for the git skills.

Usage: node github-io.ts <subcommand> --key value ...

  reply             --owner O --repo R --pr N --commentId ID --body TEXT|--body-file F
  comment           --owner O --repo R --pr N --body TEXT|--body-file F
  react             --owner O --repo R --commentId ID [--content +1]
  review            --owner O --repo R --pr N --event request-changes|comment --body TEXT|--body-file F
  resolve-thread    --threadId PRRT_...
  create-pr         --head BRANCH --base BRANCH [--title T] [--body B|--body-file F] [--draft] [--label L]
  find-run          --sha SHA
  watch-run         --runId ID
  failed-logs       --runId ID
  rerun-failed      --runId ID
  detect-workflows  (no flags — scans ./.github/workflows)

Flags are camelCase (--commentId, --threadId, --runId). The API verbs (reply,
comment, react, review) name the repository with --owner/--repo; create-pr and
the run verbs act on the repository of the CURRENT DIRECTORY (cd into its
checkout, or set GIT_SKILL_REPO) and take no --repo. Any flag a subcommand does
not list is refused. ` + "`review`" + ` cannot approve, by design.

The harness shell is zsh, which does NOT word-split unquoted $VAR — packing
flags into one variable sends them as a SINGLE argument. Use an array:
  FLAGS=(--owner o --repo r --pr 1); node github-io.ts reply "${FLAGS[@]}" --commentId 9 --body X
`

type flags map[string]string

func (o flags) req(key string) (string, error) {
	if v := o[key]; v != "" {
		return v, nil
	}
	return "", fmt.Errorf("missing required field: %s", key)
}

// buildGitHubCommand is the gh argv for a subcommand. Every required field
// is checked before any argv is returned.
func buildGitHubCommand(sub string, o flags) ([]string, error) {
	var missing error
	req := func(key string) string {
		v, err := o.req(key)
		if missing == nil {
			missing = err
		}
		return v
	}
	var argv []string
	switch sub {
	case "find-run":
		argv = []string{"run", "list", "--commit", req("sha"), "--json", "databaseId,status,conclusion,workflowName,headSha", "--limit", "20"}
	case "watch-run":
		argv = []string{"run", "watch", req("runId"), "--exit-status"}
	case "failed-logs":
		argv = []string{"run", "view", req("runId"), "--log-failed"}
	case "rerun-failed":
		argv = []string{"run", "rerun", req("runId"), "--failed"}
	case "reply":
		path := fmt.Sprintf("repos/%s/%s/pulls/%s/comments/%s/replies", req("owner"), req("repo"), req("pr"), req("commentId"))
		argv = []string{"api", "--method", "POST", path, "-f", "body=" + req("body")}
	case "resolve-thread":
		argv = []string{"api", "graphql", "-f", "query=" + resolveThreadMutation, "-f", "threadId=" + req("threadId")}
	// ── Closure for the NON-THREAD surfaces ──────────────────────────────
	// A review's summary body and a PR conversation comment have no review
	// thread: resolve-thread cannot touch them and reply does not apply.
	case "comment":
		// A PR-level comment. Quote what you answer — it lands at the bottom
		// of the conversation, not under the comment it addresses.
		path := fmt.Sprintf("repos/%s/%s/issues/%s/comments", req("owner"), req("repo"), req("pr"))
		argv = []string{"api", "--method", "POST", path, "-f", "body=" + req("body")}
	case "react":
		// An idempotent "seen/actioned" marker. GitHub exposes reactions for
		// ISSUE comments only — a review summary is closed by `comment` alone.
		path := fmt.Sprintf("repos/%s/%s/issues/comments/%s/reactions", req("owner"), req("repo"), req("commentId"))
		content := o["content"]
		if content == "" {
			content = "+1"
		}
		argv = []string{"api", "--method", "POST", path, "-f", "content=" + content}
	case "review":
		// A verdict-carrying review — what `comment` cannot be: CHANGES_REQUESTED
		// is the signal an autonomous PR author answers with a revision.
		// APPROVE IS DELIBERATELY UNREACHABLE: "never self-approve" is a hard
		// rule and --auto removes the human who enforced it, so the affordance
		// does not exist rather than existing and being discouraged.
		event, err := o.req("event")
		if err != nil {
			return nil, err
		}
		if event != "request-changes" && event != "comment" {
			return nil, fmt.Errorf("review event must be request-changes|comment (never approve): got %s", event)
		}
		argv = []string{"pr", "review", req("pr"), "--" + event, "--body", req("body"), "--repo", req("owner") + "/" + req("repo")}
	case "create-pr":
		// Ready-for-review by default; --draft is opt-in. --fill stays as the
		// fallback: gh gives an explicit --title/--body precedence over it.
		argv = []string{"pr", "create", "--head", req("head"), "--base", req("base"), "--fill"}
		if o["title"] != "" {
			argv = append(argv, "--title", o["title"])
		}
		if o["body"] != "" {
			argv = append(argv, "--body", o["body"])
		}
		if _, draft := o["draft"]; draft {
			argv = append(argv, "--draft")
		}
		// MERGE_POLICY=self repos: eve-ignore keeps eve off a PR nobody waits on.
		if o["label"] != "" {
			argv = append(argv, "--label", o["label"])
		}
	default:
		return nil, fmt.Errorf("Unknown github-io subcommand: %s", sub)
	}
	if missing != nil {
		return nil, missing
	}
	return argv, nil
}

// Workflow is one detect-workflows entry, keys in wire order.
type Workflow struct {
	Name string `json:"name"`
	Path string `json:"path"`
}

var workflowName = regexp.MustCompile(`(?m)^name:\s*(.+)$`)

// detectWorkflows lists .github/workflows/*.y{a,}ml with their `name:`;
// no workflows directory is an empty list, not an error.
func detectWorkflows(cwd string) ([]Workflow, error) {
	dir := filepath.Join(cwd, ".github", "workflows")
	entries, err := os.ReadDir(dir)
	if errors.Is(err, fs.ErrNotExist) {
		return []Workflow{}, nil
	}
	if err != nil {
		return nil, nodeFSError(err, "scandir", dir)
	}
	out := []Workflow{}
	for _, e := range entries {
		f := e.Name()
		if !strings.HasSuffix(f, ".yml") && !strings.HasSuffix(f, ".yaml") {
			continue
		}
		path := filepath.Join(dir, f)
		data, err := os.ReadFile(path)
		if err != nil {
			return nil, nodeFSError(err, "open", path)
		}
		name := f
		if m := workflowName.FindStringSubmatch(string(data)); m != nil {
			name = strings.TrimSpace(m[1])
			if len(name) > 0 && (name[0] == '"' || name[0] == '\'') {
				name = name[1:]
			}
			if len(name) > 0 && (name[len(name)-1] == '"' || name[len(name)-1] == '\'') {
				name = name[:len(name)-1]
			}
		}
		out = append(out, Workflow{Name: name, Path: ".github/workflows/" + f})
	}
	return out, nil
}

// nodeFSError renders a filesystem failure as Node's fs errors read.
func nodeFSError(err error, syscallName, path string) error {
	var errno syscall.Errno
	if errors.As(err, &errno) {
		// libuv's code and message, as Node prints them.
		codes := map[syscall.Errno][2]string{
			syscall.ENOTDIR: {"ENOTDIR", "not a directory"},
			syscall.EACCES:  {"EACCES", "permission denied"},
			syscall.EISDIR:  {"EISDIR", "illegal operation on a directory"},
			syscall.ELOOP:   {"ELOOP", "too many symbolic links encountered"},
		}
		if c, ok := codes[errno]; ok {
			return fmt.Errorf("%s: %s, %s '%s'", c[0], c[1], syscallName, path)
		}
	}
	return err
}

var whitespace = regexp.MustCompile(`\s`)

// githubIOArgs is every flag each subcommand reads (buildGitHubCommand is the
// consumer); anything else is refused. --body-file is resolved into --body
// before the build, so the build never sees it.
var githubIOArgs = map[string]verbArgs{
	"detect-workflows": {},
	"find-run":         {values: []string{"sha"}},
	"watch-run":        {values: []string{"runId"}},
	"failed-logs":      {values: []string{"runId"}},
	"rerun-failed":     {values: []string{"runId"}},
	"reply":            {values: []string{"owner", "repo", "pr", "commentId", "body", "body-file"}},
	"resolve-thread":   {values: []string{"threadId"}},
	"comment":          {values: []string{"owner", "repo", "pr", "body", "body-file"}},
	"react":            {values: []string{"owner", "repo", "commentId", "content"}},
	"review":           {values: []string{"owner", "repo", "pr", "event", "body", "body-file"}},
	"create-pr":        {values: []string{"head", "base", "title", "body", "body-file", "label"}, bools: []string{"draft"}},
}

// githubIOUsageFor is the usage line of one subcommand, for its refusals.
func githubIOUsageFor(sub string) string {
	for _, line := range strings.Split(githubIOUsage, "\n") {
		if strings.HasPrefix(line, "  "+sub+" ") {
			return "usage: vybava gitkit github-io " + strings.Join(strings.Fields(line), " ")
		}
	}
	return githubIOUsage
}

// parseFlags reads one subcommand's argv against its declaration: an unknown
// flag, a stray positional or a --draft given a value is refused, never
// dropped (a dropped --repo sent a PR to the cwd's repository, a dropped
// --body-file gave another the commit log as its body).
func parseFlags(sub string, argv []string) (flags, error) {
	// Whitespace in a flag NAME means several flags arrived glued into ONE
	// argv token — zsh not word-splitting an unquoted $VAR. Name the real
	// cause instead of a baffling `unknown argument` or `missing required
	// field`. An inline value (`--title=Fix the bug`) may hold spaces.
	for _, a := range argv {
		key, ok := strings.CutPrefix(a, "--")
		if name, _, _ := strings.Cut(key, "="); ok && whitespace.MatchString(name) {
			return nil, fmt.Errorf("flag arrived as ONE argument with embedded spaces: \"--%s\"\n"+
				"  The harness shell is zsh, which does NOT word-split unquoted $VAR.\n"+
				"  Pass the flags literally, or use an array:\n"+
				"    FLAGS=(--owner o --repo r --pr 1); node github-io.ts <sub> \"${FLAGS[@]}\"", key)
		}
	}
	spec, ok := githubIOArgs[sub]
	if !ok {
		return nil, fmt.Errorf("Unknown github-io subcommand: %s", sub)
	}
	spec.usage = githubIOUsageFor(sub)
	parsed, _, err := spec.parse("github-io "+sub, argv)
	if err != nil {
		return nil, err
	}
	return flags(parsed), nil
}

// resolveBodyFile turns --body-file into --body: the multi-line body lives in
// a file, never on a command line zsh may mangle. An empty file is refused -
// create-pr would otherwise fall back to the commit log without a word.
func resolveBodyFile(o flags) error {
	path, ok := o["body-file"]
	if !ok {
		return nil
	}
	if _, both := o["body"]; both {
		return errors.New("pass --body or --body-file, not both")
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("--body-file: %w", err)
	}
	if strings.TrimSpace(string(raw)) == "" {
		return fmt.Errorf("--body-file %s is empty", path)
	}
	delete(o, "body-file")
	o["body"] = string(raw)
	return nil
}

func runGitHubIO(args []string, stdout, stderr io.Writer) int {
	if len(args) == 0 {
		// A bare invocation exits 1, so a `$SUB` that expanded to nothing
		// fails loudly instead of looking successful.
		fmt.Fprint(stderr, githubIOUsage)
		return 1
	}
	sub, rest := args[0], args[1:]
	switch sub {
	case "help", "--help", "-h":
		fmt.Fprint(stdout, githubIOUsage)
		return 0
	case "detect-workflows":
		if _, err := parseFlags(sub, rest); err != nil {
			return fail(stderr, err)
		}
		cwd, err := syscall.Getwd()
		if err != nil {
			return fail(stderr, err)
		}
		workflows, err := detectWorkflows(cwd)
		if err != nil {
			return fail(stderr, err)
		}
		if err := writeJSON(stdout, workflows); err != nil {
			return fail(stderr, err)
		}
		return 0
	}
	o, err := parseFlags(sub, rest)
	if err != nil {
		return fail(stderr, err)
	}
	if err := resolveBodyFile(o); err != nil {
		return fail(stderr, err)
	}
	argv, err := buildGitHubCommand(sub, o)
	if err != nil {
		return fail(stderr, err)
	}
	// repoRoot(nil) — NEVER the verb's argv: github-io's own --repo is a
	// GitHub repo NAME, not a directory anchor. GIT_SKILL_REPO still applies.
	root, err := repoRoot(nil)
	if err != nil {
		return fail(stderr, err)
	}
	out, err := execFile(execOpts{dir: root, inherit: stderr, timeout: 60 * time.Second, maxBuffer: 64 << 20}, "gh", argv...)
	if err != nil {
		return fail(stderr, err)
	}
	fmt.Fprint(stdout, out)
	return 0
}
